docs/openfaas-pro/iam/function-authentication.md

```diff
@@ -0,0 +1,275 @@
+
+Authentication can be enabled on a per function basis by setting the `jwt_auth` environment variable to true in the function configuration, e.g:
+
+```yaml
+functions:
+    figlet:
+        skip_build: true
+        image: ghcr.io/openfaas/figlet:latest
+        environment:
+            jwt_auth: true
+```
+
+### Define Roles and Policies
+
+Create a new policy or update your existing policies and include the `Function:Invoke` action to allow function invocations.
+
+The following policy will allow the `env` function in the `dev` namespace to be invoked, and any function in the `openfaas-fn` namespace:
+
+```yaml
+apiVersion: iam.openfaas.com/v1
+kind: Policy
+metadata:
+  name: invoke-policy
+  namespace: openfaas
+spec:
+  statement:
+  - sid: 1-invoke-policy
+    action:
+      - "Function:Invoke"
+    effect: Allow
+    resource:
+      - "openfaas-fn:*"
+      - "dev:env"
+
+```
+
+Create a Role to map the `invoke-policy` to a user or group:
+
+```yaml
+apiVersion: iam.openfaas.com/v1
+kind: Role
+metadata:
+  name: invoke-role
+  namespace: openfaas
+spec:
+  policy:
+  - invoke-policy
+  condition:
+    principal:
+      jwt:sub:
+        - a9e0e67a-5758-4373-a4ba-23957fa66e6b
+    StringEqual:
+      jwt:iss: ["https://keycloak.example.com/realms/openfaas"]
+```
+
+
+## Invoke authenticated functions
+
+To invoke a function that has authentication enabled you need to obtain a function access token. Function access token can be obtained by exchanging a valid ID token for a function access token through the OpenFaaS Gateways token exchange endpoint.
+
+![Function authentication flow from IdP to function invocation](/images/iam/function-authentication.png)
+> Function authentication flow from IdP to function invocation.
+
+1. Obtain an ID token
+    
+    Obtain an initial ID token that can be exchanged for a function access token. This can either be a token from an identity provider registered with OpenFaaS or an OpenFaaS API access token.
+    
+    This example uses the OAuth2 client credentials flow. This flow can be used for machine to machine interactions and does not need human interactions. Any other OAuth flow or mechanism supported by your identity provider can be used to obtain the initial ID token.
+
+    === "curl"
+
+        ```bash
+        export IDP_TOKEN_URL="https://keycloak.example.com/realms/openfaas/protocol/openid-connect/token"
+        export CLIENT_ID="openfaas"
+        export CLIENT_SECRET="$(cat ./client-secret.txt)"
+
+        curl -S -L -X POST "${IDP_TOKEN_URL}" \
+            --header 'Content-Type: application/x-www-form-urlencoded' \
+            --data-urlencode "client_id=${CLIENT_ID}" \
+            --data-urlencode "client_secret=${CLIENT_SECRET}" \
+            --data-urlencode 'scope=email' \
+            --data-urlencode 'grant_type=client_credentials'
+        ```
+
+    === "ruby"
```

Comment:
I'd rather keep the ruby code in a blog post, since it sends the wrong message by omitting Python, Node, Go which are the languages most customers use.

Reply:
The ruby examples have been removed.